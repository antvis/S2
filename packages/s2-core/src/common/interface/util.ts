export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
