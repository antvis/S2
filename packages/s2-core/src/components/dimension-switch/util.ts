import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
