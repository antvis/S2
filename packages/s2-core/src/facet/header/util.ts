import { isNil } from 'lodash';

export const getCellPadding = () => {
  const padding = [12, 4, 12, 4];
  const left = isNil(padding[3]) ? 0 : padding[3];
  const right = isNil(padding[1]) ? 0 : padding[1];
  const top = isNil(padding[0]) ? 0 : padding[0];
  const bottom = isNil(padding[2]) ? 0 : padding[2];
  return {
    left,
    right,
    top,
    bottom,
  };
};
