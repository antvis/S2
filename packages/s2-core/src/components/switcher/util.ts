import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
