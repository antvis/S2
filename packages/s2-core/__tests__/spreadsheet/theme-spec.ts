/* eslint-disable jest/expect-expect */
import { createPivotSheet, createTableSheet } from 'tests/util/helpers';
import {
  CellType,
  EXTRA_COLUMN_FIELD,
  EXTRA_FIELD,
  type S2DataConfig,
  type TextAlign,
} from '@/common';
import type {
  TextBaseline,
  TextTheme,
  ThemeCfg,
  ThemeName,
} from '@/common/interface/theme';
import type { Node } from '@/facet/layout/node';
import type { PivotSheet } from '@/sheet-type';

describe('SpreadSheet Theme Tests', () => {
  let s2: PivotSheet;

  beforeEach(async () => {
    s2 = createPivotSheet(
      {
        headerActionIcons: [
          {
            icons: ['DrillDownIcon'],
            belongsCell: 'rowCell',
            displayCondition: () => true,
            onClick: () => {},
          },
        ],
      },
      { useSimpleData: false },
    );

    await s2.render();
  });

  afterEach(() => {
    s2.destroy();
  });

  describe('Theme Default Value Tests', () => {
    const CELL_TYPES: CellType[] = [
      CellType.DATA_CELL,
      CellType.ROW_CELL,
      CellType.COL_CELL,
      CellType.CORNER_CELL,
      CellType.MERGED_CELL,
    ];

    test('should get theme name', () => {
      expect(s2.getThemeName()).toEqual('default');

      s2.setThemeCfg({
        name: 'dark',
      });

      expect(s2.getThemeName()).toEqual('dark');
    });

    test('should get pivot sheet default theme', () => {
      expect(s2.theme).toMatchSnapshot();
      expect(s2.theme).toEqual(s2.getTheme());
    });

    test('should get table sheet theme', () => {
      const tableSheet = createTableSheet(null);

      expect(tableSheet.theme).toMatchSnapshot();
    });

    test.each(['dark', 'gray', 'colorful', 'default'] as ThemeName[])(
      'should get %s theme',
      (name) => {
        s2.setThemeCfg({
          name,
        });

        expect(s2.theme).toMatchSnapshot();
      },
    );

    test.each(CELL_TYPES)(
      "should assign the same color for %s's text and icon",
      async (cellType: CellType) => {
        s2.setThemeCfg({
          name: 'colorful',
        });
        await s2.render();
        const cellTheme = s2.theme[cellType];

        expect(cellTheme!.bolderText!.fill).toEqual(cellTheme!.icon!.fill);
        expect(cellTheme!.text!.fill).toEqual(cellTheme!.icon!.fill);
        expect(cellTheme!.cell).toBeTruthy();
      },
    );

    test.each(CELL_TYPES)(
      'should set cell for %s',
      async (cellType: CellType) => {
        s2.setThemeCfg({
          name: 'colorful',
        });
        await s2.render();
        const cellTheme = s2.theme[cellType];

        expect(cellTheme!.cell).toBeTruthy();
      },
    );

    test('should set theme correctly', () => {
      s2.setTheme({
        rowCell: {
          seriesNumberWidth: 200,
        },
      });

      expect(s2.theme.rowCell!.seriesNumberWidth).toStrictEqual(200);
    });

    // https://github.com/antvis/S2/issues/1538
    test('should not reset theme palette after updated theme', () => {
      const palette: ThemeCfg['palette'] = {
        basicColors: Array.from({ length: 10 }).fill('red') as string[],
        semanticColors: {
          red: 'red',
          green: 'green',
          yellow: 'yellow',
        },
      };

      s2.setThemeCfg({
        palette,
        theme: {
          rowCell: {
            text: {
              textAlign: 'left',
            },
          },
        },
      });

      s2.setTheme({
        rowCell: {
          seriesNumberWidth: 200,
        },
      });

      expect(s2.theme.background!.color).toStrictEqual('red');
    });
  });

  describe('Custom SVG Icons Tests', () => {
    test('should set custom SVG icon size and color', async () => {
      const iconInfo = {
        name: 'filter',
        size: 30,
        fill: 'red',
      };

      s2.setOptions({
        customSVGIcons: [
          {
            name: iconInfo.name,
            svg: 'https://gw.alipayobjects.com/zos/antfincdn/gu1Fsz3fw0/filter%26sort_filter.svg',
          },
        ],
        headerActionIcons: [
          {
            icons: [iconInfo.name],
            belongsCell: 'rowCell',
          },
        ],
      });
      s2.setTheme({
        rowCell: {
          icon: {
            size: iconInfo.size,
            fill: iconInfo.fill,
          },
          text: {
            fill: 'green',
          },
        },
      });

      await s2.render();

      const rowCell = s2.facet.getRowCells()[0];
      const actionIcon = rowCell.getActionIcons()[0];
      const cfg = actionIcon.getCfg();

      expect(actionIcon.name).toEqual(iconInfo.name);
      expect(cfg.fill).toEqual(iconInfo.fill);
      expect(cfg.width).toEqual(iconInfo.size);
      expect(cfg.height).toEqual(iconInfo.size);
    });
  });

  describe('Row Header Icon Tests', () => {
    const getRowCellThemeCfg = (textAlign: TextAlign) => {
      return {
        theme: {
          rowCell: {
            text: {
              textAlign,
            },
            bolderText: {
              textAlign,
            },
          },
        },
      };
    };

    const TEXT_ALIGNS: TextAlign[] = ['left', 'center', 'right'];

    test.each(TEXT_ALIGNS)(
      'should render correctly icon position when text %s aligned',
      async (align: TextAlign) => {
        s2.setThemeCfg(getRowCellThemeCfg(align));
        await s2.render();

        const rowCell = s2.facet.getRowCells()[0];

        const rowCellWidth = rowCell.getMeta().width;
        const actionIcon = rowCell.getActionIcons()[0];
        const actionIconCfg = actionIcon.getCfg();

        expect(actionIconCfg.x).toBeGreaterThanOrEqual(0);
        expect(
          (actionIconCfg.x as number) + (actionIconCfg.width as number),
        ).toBeLessThanOrEqual(rowCellWidth);
      },
    );
  });

  describe('Measure Fields Theme Tests', () => {
    const expectTextAlign = (options: {
      textAlign: TextAlign;
      fontWight: TextTheme['fontWeight'];
      containsDataCells?: boolean;
      customNodes?: Node[];
    }) => {
      const {
        textAlign,
        fontWight,
        containsDataCells = false,
        customNodes,
      } = options;
      const targetNodes = customNodes || s2.facet.getColLeafNodes();
      const dataCells = s2.facet.getDataCells();

      expect(targetNodes).not.toHaveLength(0);

      if (!containsDataCells) {
        expect(
          targetNodes.every((node) => {
            const nodeTextShape = node.belongsCell!.getTextShape();

            return (
              nodeTextShape.attr('textAlign') === textAlign &&
              nodeTextShape.attr('fontWeight') === fontWight
            );
          }),
        ).toBeTruthy();

        return;
      }

      targetNodes.forEach((node) => {
        const nodeTextShape = node.belongsCell!.getTextShape();
        const isEqualTextAlign = dataCells.every(
          (cell) =>
            cell.getTextShape().attr('textAlign') ===
            nodeTextShape.attr('textAlign'),
        );

        expect(isEqualTextAlign).toBeTruthy();
        expect(nodeTextShape.attr('fontWeight')).toStrictEqual(fontWight);
      });
      expect(dataCells[0].getTextShape().attr('textAlign')).toEqual(textAlign);
    };

    it('should default align column headers with data cells', () => {
      expectTextAlign({
        textAlign: 'right',
        fontWight: 'normal',
        containsDataCells: true,
      });
    });

    it('should render normal font wight and left text align text with row cells', async () => {
      s2.setDataCfg({
        fields: {
          valueInCols: false,
        },
      } as S2DataConfig);

      await s2.render();

      const rowMeasureFields = s2.facet
        .getRowNodes()
        .filter((node) => node.field === EXTRA_FIELD);

      expectTextAlign({
        textAlign: 'left',
        fontWight: 'normal',
        containsDataCells: false,
        customNodes: rowMeasureFields,
      });
    });

    it('should render normal font wight and left text align text with col cell', async () => {
      s2.setDataCfg({
        fields: {
          valueInCols: true,
        },
      } as S2DataConfig);

      await s2.render();

      const colMeasureFields = s2.facet
        .getColNodes()
        .filter((node) => node.field === EXTRA_FIELD);

      expectTextAlign({
        textAlign: 'right',
        fontWight: 'normal',
        containsDataCells: false,
        customNodes: colMeasureFields,
      });
    });

    it.each(['left', 'center', 'right'] as TextAlign[])(
      'should render %s text align for column nodes',
      async (textAlign) => {
        s2.setThemeCfg({
          theme: {
            colCell: {
              measureText: {
                textAlign,
              },
            },
          },
        });

        await s2.render(true);

        expectTextAlign({ textAlign, fontWight: 'normal' });
      },
    );

    it.each([
      { isRowCell: true, textAlign: 'left' },
      { isRowCell: true, textAlign: 'center' },
      { isRowCell: true, textAlign: 'right' },
      { isRowCell: false, textAlign: 'left' },
      { isRowCell: false, textAlign: 'center' },
      { isRowCell: false, textAlign: 'right' },
    ] as Array<{ isRowCell: boolean; textAlign: TextAlign }>)(
      'should render %s text align for totals nodes',
      async ({ isRowCell, textAlign }) => {
        s2.setOptions({
          totals: {
            col: {
              showGrandTotals: true,
              showSubTotals: true,
              reverseGrandTotalsLayout: true,
              reverseSubTotalsLayout: false,
            },
            row: {
              showGrandTotals: true,
              showSubTotals: true,
              reverseGrandTotalsLayout: true,
              reverseSubTotalsLayout: false,
            },
          },
        });

        s2.setThemeCfg({
          theme: {
            colCell: {
              // 小计/总计是加粗字体
              bolderText: {
                textAlign,
              },
            },
            rowCell: {
              bolderText: {
                textAlign,
              },
            },
          },
        });

        await s2.render();

        const rowTotalNodes = s2.facet
          .getRowNodes()
          .filter((node) => node.isTotals);

        const colTotalNodes = s2.facet
          .getColNodes()
          .filter((node) => node.isTotals);

        expectTextAlign({
          textAlign,
          fontWight: 700,
          customNodes: isRowCell ? rowTotalNodes : colTotalNodes,
        });
      },
    );

    it('should not align column headers with data cells and render normal font wight leaf node text if hideValue', async () => {
      s2.setOptions({
        style: {
          colCell: {
            hideValue: true,
          },
        },
        totals: null,
      });
      await s2.render();

      expectTextAlign({
        textAlign: 'center',
        fontWight: 'normal',
      });
    });

    // https://github.com/antvis/S2/pull/1371
    it.each(['left', 'center', 'right'] as TextAlign[])(
      'should render %s text align for column nodes',
      async (textAlign) => {
        s2.setThemeCfg({
          theme: {
            colCell: {
              measureText: {
                textAlign,
              },
            },
          },
        });

        s2.setDataCfg({
          fields: {
            columns: [...(s2.dataCfg.fields.columns || []), EXTRA_COLUMN_FIELD],
          },
        } as S2DataConfig);

        s2.setOptions({
          style: {
            colCell: {
              hideValue: true,
            },
          },
        });

        await s2.render(true);

        expectTextAlign({ textAlign, fontWight: 'normal' });
      },
    );
  });

  describe('Series Cell Tests', () => {
    test.each(['top', 'middle', 'bottom'] as TextBaseline[])(
      'should render %s text align for column nodes',
      async (textBaseline) => {
        s2.setThemeCfg({
          theme: {
            rowCell: {
              seriesText: {
                textBaseline,
              },
              bolderText: {
                textBaseline,
              },
            },
          },
        });

        s2.setOptions({
          seriesNumber: {
            enable: true,
          },
        });

        await s2.render();

        // 浙江省
        const rowCell = s2.facet.getRowCells()[0];
        const rowCellTextShape = rowCell.getTextShape();

        // 序号1
        const seriesCell = s2.facet.getSeriesNumberCells()[0];
        const seriesCellTextShape = seriesCell.getTextShape();

        expect(rowCellTextShape?.attr('textBaseline')).toEqual(textBaseline);
        expect(seriesCellTextShape?.attr('textBaseline')).toEqual(textBaseline);
        expect(rowCellTextShape.attr('y')).toEqual(
          seriesCellTextShape.attr('y'),
        );
      },
    );
  });

  // https://github.com/antvis/S2/issues/1892
  describe('ScrollBar Tests', () => {
    beforeEach(async () => {
      // 保证滚动条很小
      s2.setOptions({
        style: {
          rowCell: {
            width: 5000,
          },
          dataCell: {
            width: 5000,
            height: 5000,
          },
        },
      });
      await s2.render();
    });

    test('should render default min scrollbar size', () => {
      // 行头有分割线, 会减去分割线的宽度 (2px)
      expect(s2.facet.hRowScrollBar.thumbLen).toEqual(30);
      expect(s2.facet.hScrollBar.thumbLen).toEqual(32);
      expect(s2.facet.vScrollBar.thumbLen).toEqual(32);
    });

    test('should render min scrollbar size', async () => {
      s2.setTheme({
        scrollBar: {
          thumbHorizontalMinSize: 20,
          thumbVerticalMinSize: 10,
        },
      });

      await s2.render();

      // 行头有分割线, 会减去分割线的宽度 (2px)
      expect(s2.facet.hRowScrollBar.thumbLen).toEqual(18);
      expect(s2.facet.hScrollBar.thumbLen).toEqual(20);
      expect(s2.facet.vScrollBar.thumbLen).toEqual(10);
    });

    test('should render real scrollbar size', async () => {
      s2.setOptions({
        style: {
          rowCell: {
            width: 400,
          },
          dataCell: {
            width: 200,
            height: 50,
          },
        },
      });
      await s2.render();

      // 行头有分割线, 会减去分割线的宽度 (2px)
      expect(s2.facet.hRowScrollBar.thumbLen).not.toBeLessThanOrEqual(30);
      expect(s2.facet.hScrollBar.thumbLen).not.toBeLessThanOrEqual(32);
      expect(s2.facet.vScrollBar.thumbLen).not.toBeLessThanOrEqual(32);
    });
  });
});
