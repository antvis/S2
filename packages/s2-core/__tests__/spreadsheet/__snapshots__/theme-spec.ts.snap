// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpreadSheet Theme Tests Theme Default Value Tests should get default theme 1`] = `
Object {
  "background": Object {
    "color": "#FFFFFF",
    "opacity": 1,
  },
  "colCell": Object {
    "bolderText": Object {
      "fill": "#000000",
      "fontFamily": "Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif",
      "fontSize": 12,
      "fontWeight": 700,
      "opacity": 1,
      "textAlign": "center",
      "textBaseline": "middle",
    },
    "cell": Object {
      "backgroundColor": "#E0E9FD",
      "backgroundColorOpacity": 1,
      "borderDash": Array [],
      "horizontalBorderColor": "#CCDBFC",
      "horizontalBorderColorOpacity": 1,
      "horizontalBorderWidth": 1,
      "interactionState": Object {
        "highlight": Object {
          "backgroundColor": "#87B5FF",
          "backgroundOpacity": 1,
        },
        "hover": Object {
          "backgroundColor": "#CCDBFC",
          "backgroundOpacity": 0.6,
        },
        "prepareSelect": Object {
          "borderColor": "#000000",
          "borderOpacity": 1,
          "borderWidth": 1,
        },
        "searchResult": Object {
          "backgroundColor": "#F0F7FF",
          "backgroundOpacity": 1,
        },
        "selected": Object {
          "backgroundColor": "#CCDBFC",
          "backgroundOpacity": 0.6,
        },
        "unselected": Object {
          "backgroundOpacity": 0.3,
          "opacity": 0.3,
          "textOpacity": 0.3,
        },
      },
      "padding": Object {
        "bottom": 0,
        "left": 8,
        "right": 8,
        "top": 0,
      },
      "verticalBorderColor": "#CCDBFC",
      "verticalBorderColorOpacity": 1,
      "verticalBorderWidth": 1,
    },
    "icon": Object {
      "fill": "#000000",
      "margin": Object {
        "bottom": 6,
        "left": 4,
        "right": 4,
        "top": 6,
      },
      "size": 10,
    },
    "measureText": Object {
      "fill": "#000000",
      "fontFamily": "Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif",
      "fontSize": 12,
      "fontWeight": "normal",
      "opacity": 1,
      "textAlign": "right",
      "textBaseline": "middle",
    },
    "text": Object {
      "fill": "#000000",
      "fontFamily": "Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif",
      "fontSize": 12,
      "fontWeight": "normal",
      "opacity": 1,
      "textAlign": "center",
      "textBaseline": "middle",
    },
  },
  "cornerCell": Object {
    "bolderText": Object {
      "fill": "#000000",
      "fontFamily": "Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif",
      "fontSize": 12,
      "fontWeight": 700,
      "opacity": 1,
      "textAlign": "left",
      "textBaseline": "middle",
    },
    "cell": Object {
      "backgroundColor": "#E0E9FD",
      "backgroundColorOpacity": 1,
      "borderDash": Array [],
      "horizontalBorderColor": "#CCDBFC",
      "horizontalBorderColorOpacity": 1,
      "horizontalBorderWidth": 1,
      "padding": Object {
        "bottom": 0,
        "left": 8,
        "right": 8,
        "top": 0,
      },
      "verticalBorderColor": "#CCDBFC",
      "verticalBorderColorOpacity": 1,
      "verticalBorderWidth": 1,
    },
    "icon": Object {
      "fill": "#000000",
      "margin": Object {
        "left": 4,
        "right": 4,
      },
      "size": 10,
    },
    "text": Object {
      "fill": "#000000",
      "fontFamily": "Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif",
      "fontSize": 12,
      "fontWeight": 700,
      "opacity": 1,
      "textAlign": "right",
      "textBaseline": "middle",
    },
  },
  "dataCell": Object {
    "bolderText": Object {
      "fill": "#000000",
      "fontFamily": "Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif",
      "fontSize": 12,
      "fontWeight": 700,
      "linkTextFill": "#326EF4",
      "opacity": 1,
      "textAlign": "right",
      "textBaseline": "middle",
    },
    "cell": Object {
      "backgroundColor": "#FFFFFF",
      "backgroundColorOpacity": 1,
      "borderDash": Array [],
      "crossBackgroundColor": "#F5F8FE",
      "horizontalBorderColor": "#E0E9FD",
      "horizontalBorderColorOpacity": 1,
      "horizontalBorderWidth": 1,
      "interactionState": Object {
        "highlight": Object {
          "backgroundColor": "#87B5FF",
          "backgroundOpacity": 1,
        },
        "hover": Object {
          "backgroundColor": "#E0E9FD",
          "backgroundOpacity": 0.6,
        },
        "hoverFocus": Object {
          "backgroundColor": "#E0E9FD",
          "backgroundOpacity": 0.6,
          "borderColor": "#000000",
          "borderOpacity": 1,
          "borderWidth": 1,
        },
        "prepareSelect": Object {
          "borderColor": "#000000",
          "borderOpacity": 1,
          "borderWidth": 1,
        },
        "searchResult": Object {
          "backgroundColor": "#F0F7FF",
          "backgroundOpacity": 1,
        },
        "selected": Object {
          "backgroundColor": "#E0E9FD",
          "backgroundOpacity": 0.6,
        },
        "unselected": Object {
          "backgroundOpacity": 0.3,
          "opacity": 0.3,
          "textOpacity": 0.3,
        },
      },
      "padding": Object {
        "bottom": 8,
        "left": 8,
        "right": 8,
        "top": 8,
      },
      "verticalBorderColor": "#E0E9FD",
      "verticalBorderColorOpacity": 1,
      "verticalBorderWidth": 1,
    },
    "icon": Object {
      "downIconColor": "#FF4D4F",
      "fill": "#000000",
      "margin": Object {
        "left": 4,
        "right": 4,
      },
      "size": 10,
      "upIconColor": "#29A294",
    },
    "miniChart": Object {
      "bar": Object {
        "fill": "#326EF4",
        "intervalPadding": 4,
        "opacity": 1,
      },
      "bullet": Object {
        "backgroundColor": "#E9E9E9",
        "comparativeMeasure": Object {
          "fill": "#000000",
          "height": 12,
          "opacity": 0.25,
          "width": 1,
        },
        "progressBar": Object {
          "height": 10,
          "innerHeight": 6,
          "widthPercent": 0.6,
        },
        "rangeColors": Object {
          "bad": "#FF4D4F",
          "good": "#29A294",
          "satisfactory": "#FAAD14",
        },
      },
      "interval": Object {
        "fill": "#326EF4",
        "height": 12,
      },
      "line": Object {
        "linkLine": Object {
          "fill": "#326EF4",
          "opacity": 0.6,
          "size": 1.5,
        },
        "point": Object {
          "fill": "#326EF4",
          "opacity": 1,
          "size": 2.2,
        },
      },
    },
    "text": Object {
      "fill": "#000000",
      "fontFamily": "Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif",
      "fontSize": 12,
      "fontWeight": "normal",
      "linkTextFill": "#326EF4",
      "opacity": 1,
      "textAlign": "right",
      "textBaseline": "middle",
    },
  },
  "mergedCell": Object {
    "bolderText": Object {
      "fill": "#000000",
      "fontFamily": "Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif",
      "fontSize": 12,
      "fontWeight": 700,
      "linkTextFill": "#326EF4",
      "opacity": 1,
      "textAlign": "right",
      "textBaseline": "middle",
    },
    "cell": Object {
      "backgroundColor": "#FFFFFF",
      "backgroundColorOpacity": 1,
      "borderDash": Array [],
      "crossBackgroundColor": "#F5F8FE",
      "horizontalBorderColor": "#E0E9FD",
      "horizontalBorderColorOpacity": 1,
      "horizontalBorderWidth": 1,
      "interactionState": Object {
        "highlight": Object {
          "backgroundColor": "#87B5FF",
          "backgroundOpacity": 1,
        },
        "hover": Object {
          "backgroundColor": "#E0E9FD",
          "backgroundOpacity": 0.6,
        },
        "hoverFocus": Object {
          "backgroundColor": "#E0E9FD",
          "backgroundOpacity": 0.6,
          "borderColor": "#000000",
          "borderOpacity": 1,
          "borderWidth": 1,
        },
        "prepareSelect": Object {
          "borderColor": "#000000",
          "borderOpacity": 1,
          "borderWidth": 1,
        },
        "searchResult": Object {
          "backgroundColor": "#F0F7FF",
          "backgroundOpacity": 1,
        },
        "selected": Object {
          "backgroundColor": "#E0E9FD",
          "backgroundOpacity": 0.6,
        },
        "unselected": Object {
          "backgroundOpacity": 0.3,
          "opacity": 0.3,
          "textOpacity": 0.3,
        },
      },
      "padding": Object {
        "bottom": 8,
        "left": 8,
        "right": 8,
        "top": 8,
      },
      "verticalBorderColor": "#E0E9FD",
      "verticalBorderColorOpacity": 1,
      "verticalBorderWidth": 1,
    },
    "icon": Object {
      "downIconColor": "#FF4D4F",
      "fill": "#000000",
      "margin": Object {
        "left": 4,
        "right": 4,
      },
      "size": 10,
      "upIconColor": "#29A294",
    },
    "miniChart": Object {
      "bar": Object {
        "fill": "#326EF4",
        "intervalPadding": 4,
        "opacity": 1,
      },
      "bullet": Object {
        "backgroundColor": "#E9E9E9",
        "comparativeMeasure": Object {
          "fill": "#000000",
          "height": 12,
          "opacity": 0.25,
          "width": 1,
        },
        "progressBar": Object {
          "height": 10,
          "innerHeight": 6,
          "widthPercent": 0.6,
        },
        "rangeColors": Object {
          "bad": "#FF4D4F",
          "good": "#29A294",
          "satisfactory": "#FAAD14",
        },
      },
      "interval": Object {
        "fill": "#326EF4",
        "height": 12,
      },
      "line": Object {
        "linkLine": Object {
          "fill": "#326EF4",
          "opacity": 0.6,
          "size": 1.5,
        },
        "point": Object {
          "fill": "#326EF4",
          "opacity": 1,
          "size": 2.2,
        },
      },
    },
    "text": Object {
      "fill": "#000000",
      "fontFamily": "Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif",
      "fontSize": 12,
      "fontWeight": "normal",
      "linkTextFill": "#326EF4",
      "opacity": 1,
      "textAlign": "right",
      "textBaseline": "middle",
    },
  },
  "prepareSelectMask": Object {
    "backgroundColor": "#234DAB",
    "backgroundOpacity": 0.3,
  },
  "resizeArea": Object {
    "background": "#326EF4",
    "backgroundOpacity": 0,
    "guideLineColor": "#326EF4",
    "guideLineDash": Array [
      3,
      3,
    ],
    "guideLineDisableColor": "rgba(0,0,0,0.25)",
    "interactionState": Object {
      "hover": Object {
        "backgroundColor": "#326EF4",
        "backgroundOpacity": 1,
      },
    },
    "size": 3,
  },
  "rowCell": Object {
    "bolderText": Object {
      "fill": "#000000",
      "fontFamily": "Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif",
      "fontSize": 12,
      "fontWeight": 700,
      "linkTextFill": "#326EF4",
      "opacity": 1,
      "textAlign": "left",
      "textBaseline": "top",
    },
    "cell": Object {
      "backgroundColor": "#F5F8FE",
      "backgroundColorOpacity": 1,
      "borderDash": Array [],
      "horizontalBorderColor": "#E0E9FD",
      "horizontalBorderColorOpacity": 1,
      "horizontalBorderWidth": 1,
      "interactionState": Object {
        "highlight": Object {
          "backgroundColor": "#87B5FF",
          "backgroundOpacity": 1,
        },
        "hover": Object {
          "backgroundColor": "#E0E9FD",
          "backgroundOpacity": 0.6,
        },
        "prepareSelect": Object {
          "borderColor": "#000000",
          "borderOpacity": 1,
          "borderWidth": 1,
        },
        "searchResult": Object {
          "backgroundColor": "#F0F7FF",
          "backgroundOpacity": 1,
        },
        "selected": Object {
          "backgroundColor": "#E0E9FD",
          "backgroundOpacity": 0.6,
        },
        "unselected": Object {
          "backgroundOpacity": 0.3,
          "opacity": 0.3,
          "textOpacity": 0.3,
        },
      },
      "padding": Object {
        "bottom": 0,
        "left": 8,
        "right": 8,
        "top": 0,
      },
      "verticalBorderColor": "#E0E9FD",
      "verticalBorderColorOpacity": 1,
      "verticalBorderWidth": 1,
    },
    "icon": Object {
      "fill": "#000000",
      "margin": Object {
        "left": 4,
        "right": 4,
      },
      "size": 10,
    },
    "measureText": Object {
      "fill": "#000000",
      "fontFamily": "Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif",
      "fontSize": 12,
      "fontWeight": "normal",
      "linkTextFill": "#326EF4",
      "opacity": 1,
      "textAlign": "left",
      "textBaseline": "top",
    },
    "seriesNumberWidth": 80,
    "seriesText": Object {
      "fill": "#000000",
      "fontFamily": "Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif",
      "fontSize": 12,
      "fontWeight": "normal",
      "linkTextFill": "#326EF4",
      "opacity": 1,
      "textAlign": "center",
      "textBaseline": "top",
    },
    "text": Object {
      "fill": "#000000",
      "fontFamily": "Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif",
      "fontSize": 12,
      "fontWeight": "normal",
      "linkTextFill": "#326EF4",
      "opacity": 1,
      "textAlign": "left",
      "textBaseline": "top",
    },
  },
  "scrollBar": Object {
    "hoverSize": 8,
    "lineCap": "round",
    "size": 6,
    "thumbColor": "rgba(0,0,0,0.15)",
    "thumbHorizontalMinSize": 32,
    "thumbHoverColor": "rgba(0,0,0,0.25)",
    "thumbVerticalMinSize": 32,
    "trackColor": "rgba(0,0,0,0.01)",
  },
  "splitLine": Object {
    "borderDash": Array [],
    "horizontalBorderColor": "#326EF4",
    "horizontalBorderColorOpacity": 0.2,
    "horizontalBorderWidth": 2,
    "shadowColors": Object {
      "left": "rgba(0,0,0,0.1)",
      "right": "rgba(0,0,0,0)",
    },
    "shadowWidth": 8,
    "showShadow": true,
    "verticalBorderColor": "#326EF4",
    "verticalBorderColorOpacity": 0.25,
    "verticalBorderWidth": 2,
  },
}
`;
